pal/AzureIoTSocket_WiFi/socketio_esp32wifi.cpp

```diff
@@ -0,0 +1,554 @@
+            if (first_pending_io != NULL)
+            {
+                if (add_pending_io(socket_io_instance, (const unsigned char*)buffer, size, on_send_complete, callback_context) != 0)
+                {
+                    LogError("Failure: add_pending_io failed.");
+                    result = MU_FAILURE;
+                }
+                else
+                {
+                    result = 0;
+                }
+            }
+            else
+            {
+                ssize_t send_result = socket_io_instance->socket->write((const uint8_t *)buffer, size);
+
+                if (send_result == 0)
+                {
+                    LogError("Failure: sending socket failed. errno=%d (%s).", errno, strerror(errno));
+                    result = MU_FAILURE;
+                }
+                else if (send_result < size)
+                {
+                    /* queue data */
+                    if (add_pending_io(socket_io_instance, (const unsigned char*)(buffer + send_result), size - send_result, on_send_complete, callback_context) != 0)
+                    {
+                        LogError("Failure: add_pending_io failed.");
+                        result = MU_FAILURE;
+                    }
+                    else
+                    {
+                        result = send_result;
+                    }
+                }
+                else
+                {
+                    if (on_send_complete != NULL)
+                    {
+                        on_send_complete(callback_context, IO_SEND_OK);
+                    }
+
+                    result = 0;
+                }
+            }
+        }
+    }
+
+    return result;
+}
+
+void socketio_dowork(CONCRETE_IO_HANDLE socket_io)
+{
+    if (socket_io != NULL)
+    {
+        SOCKET_IO_INSTANCE* socket_io_instance = (SOCKET_IO_INSTANCE*)socket_io;
+        LIST_ITEM_HANDLE first_pending_io = singlylinkedlist_get_head_item(socket_io_instance->pending_io_list);
+        while (first_pending_io != NULL)
+        {
+            PENDING_SOCKET_IO* pending_socket_io = (PENDING_SOCKET_IO*)singlylinkedlist_item_get_value(first_pending_io);
+            if (pending_socket_io == NULL)
+            {
+                indicate_error(socket_io_instance);
+                LogError("Failure: retrieving socket from list");
+                break;
+            }
+
+            ssize_t send_result = socket_io_instance->socket->write(pending_socket_io->bytes, pending_socket_io->size);
+            if (send_result == 0)
+            {
+                free(pending_socket_io->bytes);
+                free(pending_socket_io);
+                (void)singlylinkedlist_remove(socket_io_instance->pending_io_list, first_pending_io);
+
+                LogError("Failure: sending Socket information");
+                indicate_error(socket_io_instance);
+            }
```

Comment:
can you check with @ewertons how to handle EAGAIN?

Reply:
After looking into this, EAGAIN is not an issue here as EAGAIN is handled internally by the board's wifi package when we call write().